### ISSUE TITLE
LibGit2 segfault on exit

### ISSUE THREAD

**Issue by simonbyrne** [JuliaLang/julia#20109]
I seem to be getting segfaults when Julia exits, and there are LibGit2 finalizers to be run (probably due to #19660).

e.g. from within current Julia git directory, this reliably causes a segfault on my machine:
```
├ juliadev -e 'r = LibGit2.GitRepo("."); LibGit2.get(LibGit2.GitCommit, r, LibGit2.GitHash("48e923fb1ed7c27350441bf36696abf66a015716"));'

signal (11): Segmentation fault: 11
while loading no file, in expression starting on line 0
git_mwindow_put_pack at /Users/simon/src/julia/usr/lib/libgit2.0.25.0.dylib (unknown line)
pack_backend__free at /Users/simon/src/julia/usr/lib/libgit2.0.25.0.dylib (unknown line)
git_odb_free at /Users/simon/src/julia/usr/lib/libgit2.0.25.0.dylib (unknown line)
git_repository__cleanup at /Users/simon/src/julia/usr/lib/libgit2.0.25.0.dylib (unknown line)
git_repository_free at /Users/simon/src/julia/usr/lib/libgit2.0.25.0.dylib (unknown line)
close at ./libgit2/types.jl:491
jlcall_close_23653 at /Users/simon/src/julia/usr/lib/julia/sys.dylib (unknown line)
jl_apply_generic at /Users/simon/src/julia/src/gf.c:2214
jl_apply at /Users/simon/src/julia/src/./julia.h:1413 [inlined]
run_finalizer at /Users/simon/src/julia/src/gc.c:110
jl_gc_run_finalizers_in_list at /Users/simon/src/julia/src/gc.c:193 [inlined]
run_finalizers at /Users/simon/src/julia/src/gc.c:221
jl_atexit_hook at /Users/simon/src/julia/src/init.c:223
main at /Users/simon/bin/juliadev (unknown line)
Allocations: 1007473 (Pool: 1006322; Big: 1151); GC: 0
Segmentation fault: 11
```

**Comment by simonbyrne** [comment]
cc: @tkelman

**Comment by tkelman** [comment]
yes, just bisected to https://github.com/JuliaLang/julia/pull/19660 from running `Pkg.test("PkgBenchmark")`

**Comment by simonbyrne** [comment]
It's kind of weird. It goes away if I manually close the objects before exiting.

**Comment by simonbyrne** [comment]
And I still see the problem if I revert to libgit2 v0.24.5

**Comment by simonbyrne** [comment]
@tkelman do you only see the segfault at exit as well?

**Comment by tkelman** [comment]
So far, yeah

**Comment by simonbyrne** [comment]
Yeah, it seems to be something funny happening at exit. e.g.
```
julia> function foo()
       r = LibGit2.GitRepo(".")
       c = LibGit2.get(LibGit2.GitCommit, r, LibGit2.GitHash("48e923fb1ed7c27350441bf36696abf66a015716"))
       return true
       end
foo (generic function with 1 method)

julia> foo()
true

julia> gc()
```
works fine. But if I leave off the `gc()` call, I get the segfault.

**Comment by simonbyrne** [comment]
It goes away if I disable [this `atexit` hook](https://github.com/JuliaLang/julia/blob/264872d260057e7b07e32ff82e1081c00c6760e2/base/libgit2/libgit2.jl#L606-L608)

**Comment by simonbyrne** [comment]
Yup, that's it:
```
julia> r = LibGit2.GitRepo(".");

julia> c = LibGit2.get(LibGit2.GitCommit, r, LibGit2.GitHash("48e923fb1ed7c27350441bf36696abf66a015716"));

julia> ccall((:git_libgit2_shutdown, :libgit2), Cint, ())
0

julia> close(r)

signal (11): Segmentation fault: 11
while loading no file, in expression starting on line 0
git_mwindow_put_pack at /Users/simon/src/julia/usr/lib/libgit2.0.24.5.dylib (unknown line)
pack_backend__free at /Users/simon/src/julia/usr/lib/libgit2.0.24.5.dylib (unknown line)
git_odb_free at /Users/simon/src/julia/usr/lib/libgit2.0.24.5.dylib (unknown line)
git_repository__cleanup at /Users/simon/src/julia/usr/lib/libgit2.0.24.5.dylib (unknown line)
git_repository_free at /Users/simon/src/julia/usr/lib/libgit2.0.24.5.dylib (unknown line)
close at ./libgit2/types.jl:466
jlcall_close_24615 at /Users/simon/src/julia/usr/lib/julia/sys.dylib (unknown line)
jl_apply_generic at /Users/simon/src/julia/src/gf.c:2214
do_call at /Users/simon/src/julia/src/interpreter.c:75
eval at /Users/simon/src/julia/src/interpreter.c:215
jl_interpret_toplevel_expr at /Users/simon/src/julia/src/interpreter.c:34
jl_toplevel_eval_flex at /Users/simon/src/julia/src/toplevel.c:640
jl_toplevel_eval_in at /Users/simon/src/julia/src/builtins.c:614
eval at ./boot.jl:238
jlcall_eval_18808 at /Users/simon/src/julia/usr/lib/julia/sys.dylib (unknown line)
jl_apply_generic at /Users/simon/src/julia/src/gf.c:2214
eval_user_input at ./REPL.jl:66
unknown function (ip: 0x3119c9696)
jl_apply_generic at /Users/simon/src/julia/src/gf.c:2214
macro expansion at ./REPL.jl:97 [inlined]
#1 at ./event.jl:71
unknown function (ip: 0x3119c093f)
jl_apply_generic at /Users/simon/src/julia/src/gf.c:2214
jl_apply at /Users/simon/src/julia/src/./julia.h:1413 [inlined]
start_task at /Users/simon/src/julia/src/task.c:261
Allocations: 3162223 (Pool: 3160897; Big: 1326); GC: 3
Segmentation fault: 11
```

**Comment by simonbyrne** [comment]
@yuyichao You mentioned [in this discussion](https://discourse.julialang.org/t/finalizer-and-atexit-for-c-library-wrappers/772) that it might be possible to add post-finalize atexit functions? How feasible would this be?

**Comment by vtjnash** [comment]
Does that function also cleanup all of the libgit2 objects? I think you could just set a flag before calling shutdown, and return early from any finalizers afterwards

**Comment by simonbyrne** [comment]
No it doesn't, all objects need to be cleaned up before calling `git_libgit2_shutdown`.

**Comment by yuyichao** [comment]
> add post-finalize atexit functions?

That can be hard to specify what you can do and what you can't in it so I'd rather not unless it's really necessary.

I think you should just refcount the global context in the libgit2 object constructor and finalizers.

**Comment by simonbyrne** [comment]
Should be fixed by #20124.

**Comment by tkelman** [comment]
Confirming that `Pkg.test("PkgBenchmark")` works fine now.
